File: Example/Assets/OpenPath/OPScanner.js
#pragma strict

class OPScanner extends MonoBehaviour {
	public var scanOnEnable : boolean = true;
	public var layerMask : LayerMask;
	public var mapType : OPMapType;
	public var map : OPMap = null;
	public var heuristic : float = 10.0;
	public var spacing : float = 1.0;
	
	private var bounds : Bounds;
	private var gridSize : Vector3;
	
	public function GenerateBounds () {
	    var bounds : Bounds = new Bounds ( Vector3.zero, Vector3.zero );
	    var pos : Vector3 = Vector3.zero;
	    
	    for ( var r : Renderer in FindObjectsOfType ( Renderer ) ) {
	    	bounds.Encapsulate ( r.bounds );
	    }
	    
	    pos = bounds.min;
	    
	    transform.position = pos;
		gridSize = new Vector3 ( Mathf.Round ( bounds.size.x / spacing ) + 1, Mathf.Round ( bounds.size.y / spacing ) + 1, Mathf.Round ( bounds.size.z / spacing ) + 1 );
	}
	
	public function SetMap ( nodes : OPNode[] ) {
		map = new OPMap ();
		map.nodes = nodes;
	}
	
	public function GenerateMap () {
		if ( mapType == OPMapType.Grid ) {
			map = new OPGridMap ( transform.position, gridSize, spacing, layerMask );
		
		} else if ( mapType == OPMapType.NavMesh ) {
			var navMesh : OPNavMesh = GameObject.FindObjectOfType ( OPNavMesh );
			
			map = new OPNavMeshMap ( navMesh );
		
		} else {
			map = new OPWayPointMap ( GameObject.FindObjectsOfType ( OPWayPoint ) );
		
		}
	}

	public function Clear () {
		map = null;

		this.transform.position = Vector3.zero;

		Debug.Log ( "OPScanner | Cleared nodes" );
	}

	public function Scan () {
		Clear ();
		
		Debug.Log ( "OPScanner | Scanning for navigation nodes as " + mapType + "..." );
		
		GenerateBounds ();
		GenerateMap ();
		
		Debug.Log ( "OPScanner | ...scan completed" );
	}
	
	function Start () {
		if ( scanOnEnable ) {
			Scan ();
		}
	}
	
	public function FindPath ( start : Vector3, goal : Vector3 ) : List.<OPNode> {
		var here : OPNode = GetClosestNode ( start );
		var there : OPNode = GetClosestNode ( goal );
		var list : List.<OPNode> = OPAStar.Search ( here, there, map, heuristic );
	
		map.Reset ();
	
		return list;
	}
	
	function OnDrawGizmos () {
		if ( map == null ) { return; }
		if ( map.nodes == null ) { return; }
		
		if ( bounds != null ) {
			Gizmos.color = Color.white;
		
			Gizmos.DrawWireCube ( bounds.center, bounds.size );
		}
		
		for ( var n : OPNode in map.nodes ) {		
			if ( n == null ) { continue; }
			
			Gizmos.color = new Color ( 0, 0.8, 1, 1 );
			
			if ( n.parent ) { Gizmos.color = Color.red; }
			if ( n.active ) { Gizmos.color = Color.green; }
			if ( n.neighbors.Count < 1 ) { Gizmos.color = Color.red; }
			
			Gizmos.DrawCube ( n.position, new Vector3 ( 0.1, 0.1, 0.1 ) );
			
			Gizmos.color = Color.green;
			
			for ( var nb : OPNode in n.neighbors ) {
				if ( n.active && nb.active ) {
					Gizmos.DrawLine ( n.position, nb.position );
				}
			}
			
			Gizmos.color = Color.white;
		}
	}
	
	public function GetClosestNode ( pos : Vector3 ) : OPNode {
		var shortestDistance : float = 100;
		var node : OPNode;
		
		for ( var n : OPNode in map.nodes ) {
			if ( n == null ) { continue; }
			
			var currentDistance : float = ( pos - (n as OPNode).position ).magnitude;
			
			if ( currentDistance < shortestDistance && ( n as OPNode ).neighbors.Count > 0 ) {
				shortestDistance = currentDistance;
				node = n as OPNode;
				
			}
		}
		
		return node;
	}
}

File: C#/Example/Assets/OpenPath/OPScanner.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class OPScanner : MonoBehaviour {

    public bool scanOnEnable = true;
    public LayerMask layerMask;
    public Vector3 gridSize;
    public OPMapType mapType;
    public OPMap map = null;
    public float heuristic = 10f;
    public float spacing = 1f;
    public Bounds bounds;

    void Start()
    {
        if (scanOnEnable) { Scan(); }
    }

    void GetBounds()
    {
        Bounds bounds = new Bounds(Vector3.zero, Vector3.zero);
        Vector3 pos = Vector3.zero;

        foreach (Renderer r in FindObjectsOfType<Renderer>())
        {
            bounds.Encapsulate(r.bounds);
        }

        pos = bounds.min;

        transform.position = pos;
        gridSize = new Vector3(Mathf.Round(bounds.size.x / spacing) + 1, Mathf.Round(bounds.size.y / spacing) + 1, Mathf.Round(bounds.size.z / spacing) + 1);
    }

    void SetMap(OPNode[] nodes)
    {
        map = new OPMap();
        map.nodes = nodes;
    }

    void SetMap()
    {
        if (mapType == OPMapType.Grid)
        {
            map = new OPGridMap(transform.position, gridSize, spacing, layerMask);
        }
        else if ( mapType == OPMapType.NavMesh ) {
			OPNavMesh navMesh = GameObject.FindObjectOfType <OPNavMesh>( );
			
			map = new OPNavMeshMap ( navMesh );
		
		} 
        else {
			map = new OPWayPointMap ( GameObject.FindObjectsOfType<OPWayPoint>());
		}
    }

    IEnumerator Init()
    {
        Debug.Log("OPScanner | Scanning for navigation nodes as " + mapType + "...");

        float timeTaken = Time.time;

        yield return new WaitForEndOfFrame();

        GetBounds();
        SetMap();

        yield return new WaitForEndOfFrame();

        timeTaken = (Time.time - timeTaken) * 10;

        Debug.Log("OPScanner | ... scane complete in " + timeTaken + " seconds");
    }

    void Scan()
    {
        StartCoroutine(Init());
    }

    public List<OPNode> FindPath(Vector3 start, Vector3 goal)
    {
        OPNode here = GetClosestNode(start);
        OPNode there = GetClosestNode(goal);
        List<OPNode> list = OPAStar.Search(here, there, map, heuristic);

        map.Reset();

        return list;
    }

    void OnDrawGizmos()
    {
        if (map == null) { return; }
        if (map.nodes == null) { return; }

        if (bounds != null)
        {
            Gizmos.color = Color.white;
            Gizmos.DrawWireCube(bounds.center, bounds.size);
        }

        foreach (OPNode n in map.nodes)
        {
            if (n == null) { continue; }

            Gizmos.color = new Color(1f, 1f, 1f, 0.5f);

            if (n.parent != null) { Gizmos.color = Color.red; }
            if (n.active) { Gizmos.color = Color.green; }
            if (n.neighbors.Count < 1) { Gizmos.color = Color.red; }

            Gizmos.DrawCube(n.position, new Vector3(0.1f, 0.1f, 0.1f));

            Gizmos.color = Color.green;

            foreach (OPNode nb in n.neighbors)
            {
                if (n.active && nb.active)
                {
                    Gizmos.DrawLine(n.position, nb.position);
                }
            }

            Gizmos.color = Color.white;
        }
    }

    OPNode GetClosestNode(Vector3 pos)
    {
        float shortestDistance = 100f;
        OPNode node = new OPNode();
        //print("Length " + map.nodes.Length);

        foreach (OPNode n in map.nodes)
        {
            if ( n == null ) { continue; }
			
			float currentDistance = ( pos - (n as OPNode).position ).magnitude;
            //print(currentDistance);
			
			if ( currentDistance < shortestDistance && ( n as OPNode ).neighbors.Count > 0 ) {
				shortestDistance = currentDistance;
				node = n as OPNode;
				
			}
        }
        //print(node.position);
        return node;
    }
}
